File: 2197. Replace Non-Coprime Numbers in Array.cpp
class Solution {
public:
    int gcdComp(int a, int b) {
        if (a == 0) return b;
        return gcdComp(b % a, a);
    }

    vector<int> replaceNonCoprimes(vector<int>& nums) {
        stack<int> st;
        vector<int> res;

        for (int num : nums) {
            long long x = num; 
            while (!st.empty()) {
                int top = st.top();
                int g = gcdComp(top, (int)x);
                if (g == 1) break;

                x = x / g * top;
                st.pop();
            }
            st.push((int)x);
        }

        while (!st.empty()) {
            res.push_back(st.top());
            st.pop();
        }
        reverse(res.begin(), res.end());
        return res;
    }
};

File: 2197. Replace Non-Coprime Numbers in Array.java
class Solution {
    public int gcdComp(int a, int b){
        if(a == 0) return b;
        return gcdComp(b % a, a);
    }
    public List<Integer> replaceNonCoprimes(int[] nums) {
        Stack<Integer> st = new Stack<>();
        List<Integer> res = new ArrayList<>();
        
        for(int num : nums){
            int x = num;
            while(!st.isEmpty()){
                int top = st.peek();
                int gcd = gcdComp(top, x);
                if(gcd == 1) break;
            
                x /= gcd;
                x *= top;
                st.pop();
            }
            st.push(x);
        }

        while(!st.isEmpty()) res.add(st.pop());
        Collections.reverse(res);
        return res;
    }
}